Add new variable 'mysql-query_rules_fast_routing_algorithm'

JavierJF opened the pull request:
This PR introduces this two big changes:

 - Implemented new variable 'query_rules_fast_routing_algorithm' that allows choosing between the classic per-thread hashmaps for 'query_rules_fast_routing' searches (1), and a new unique global hashmap (2).
 - Improved locking time on 'QueryProcessor::rwlock' during 'load_mysql_query_rules_to_runtime'.

And several smaller ones.

JavierJF committed:
Add new variable 'mysql-query_rules_fast_routing_algorithm'

 - Implemented new variable 'query_rules_fast_routing_algorithm' that
   allows choosing between the classic per-thread hashmaps for
   'query_rules_fast_routing' searches (1), and a new unique global
   hashmap (2).
 - Improved locking time on 'QueryProcessor::rwlock' during
   'load_mysql_query_rules_to_runtime'.
 - Changed computation of 'mysql_query_rules_memory' to be lazily
   updated by threads themselves while building the per-threads maps.
 - Fixed 'ProxySQL_Admin::stats___memory_metrics' report of the memory
   metrics.
 - Improved 'PROXYSQLTEST 17' command, for being able to micro-benchmark
   and check 'query_rules_fast_routing_algorithm'.

JavierJF committed:
Add testing for 'mysql-query_rules_fast_routing_algorithm'

- Improved test 'test_sqlite3_server_and_fast_routing-t' for performing
  the same checks with both supported values for the variable.
- Added new test 'test_query_rules_fast_routing_algorithm-t'

JavierJF committed:
Remove thread variable for 'mysql-query_rules_fast_routing_algorithm'

 - Thread variable has been replaced for new variable in 'Query_Processor'.
 - Computation for 'mysql_query_rules_memory' is now performed by Admin
   during 'load_mysql_query_rules_to_runtime' instead of lazily by the
   MySQL_Threads.

JavierJF committed:
Improve 'test_query_rules_fast_routing_algorithm-t'

 - Added new utility functions for checking 'debugdb' written entries.
 - Test now performs simultaneous checks on 'logfile' and 'debugdb'.
 - Moved lambda functions into proper functions.

JavierJF committed:
Merge branch 'v2.x' of github.com:sysown/proxysql into v2.x-query_rules_fast_routing_algorithm

JavierJF committed:
Cleanup of test 'test_query_rules_fast_routing_algorithm-t' utilities

 - Moved new tap utility functions to 'utils.h'.
 - Removed previously borrowed utilities from another PR.

JavierJF committed:
Add missing linker flag for 'sqlite3' for tap 'tests_with_deps'

renecannao commented on lib/MySQL_Thread.cpp:
```diff
@@ -1666,7 +1668,14 @@ bool MySQL_Threads_Handler::set_variable(char *name, const char *value) {	// thi
+				if (GloQPro) {
+					int intv = atoi(value);
+					if (intv >= std::get<1>(it->second) && intv <= std::get<2>(it->second)) {
+						GloQPro->query_rules_fast_routing_algorithm = intv;
```
Should we lock the Query Processor first?

JavierJF replied:
Indeed, somehow I forgot to place this look, even If I brought up the topic in the previous discussion :facepalm: . Thanks for the heads up, it's now fixed.

renecannao commented on lib/Query_Processor.cpp:
```diff
@@ -431,7 +431,7 @@ static void __delete_query_rule(QP_rule_t *qr) {
 // delete all the query rules in a Query Processor Table
 // Note that this function is called by GloQPro with &rules (generic table)
 //     and is called by each mysql thread with _thr_SQP_rules (per thread table)
```
This comment needs correction. 
Now `__reset_rules()` is also called from Admin .
We also need to describe why

JavierJF replied:
Now properly commented.

renecannao commented on lib/Query_Processor.cpp:
```diff
@@ -1993,30 +2053,23 @@ Query_Processor_Output * Query_Processor::process_mysql_query(MySQL_Session *ses
+		} else if (rules_fast_routing != nullptr) {
+			pthread_rwlock_rdlock(&this->rwlock);
+			proxy_debug(PROXY_DEBUG_MYSQL_QUERY_PROCESSOR, 7, "Searching global 'rules_fast_routing' hashmap with: user='%s', schema='%s', and flagIN='%d'\n", u, s, flagIN);
+			dst_hg = search_rules_fast_routing_dest_hg(rules_fast_routing, u, s, flagIN);
```
Micro optimization.
Because `search_rules_fast_routing_dest_hg()` performs a series of `strlen()`, `sprintf()` , optionally `malloc()` and `free()` , the locking should be done inside `search_rules_fast_routing_dest_hg()` .
We can simply pass a `bool` to `search_rules_fast_routing_dest_hg()` to either perform the lock or not.

JavierJF replied:
I have added several optimizations, including the suggested ones.

JavierJF committed:
Lock 'Query_Processor' for 'query_rules_fast_routing_algorithm' variable update

JavierJF committed:
Optimize locking for 'search_rules_fast_routing_dest_hg'

 - Add parameter for conditionally locking inside function.
 - Move potential syscalls outside the 'pthread_rwlock_rdlock' section.
 - Reduce number of 'sprintf' calls.
 - Improved PROXYSQLTEST 11/15 with data randomly exceeding function
   default buffer size.

JavierJF committed:
Fix outdated comment for '__reset_rules' function
